source/AsyncRgbLedSimulationDataGenerator.cpp

```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
+	mLEDSimulationData.SetInitialBitState( BIT_HIGH );
 }
 
 U32 AsyncRgbLedSimulationDataGenerator::GenerateSimulationData( U64 largest_sample_requested, U32 sample_rate, SimulationChannelDescriptor** simulation_channel )
 {
-	U64 adjusted_largest_sample_requested = AnalyzerHelpers::AdjustSimulationTargetSample( largest_sample_requested, sample_rate, mSimulationSampleRateHz );
+	U64 adjusted_largest_sample_requested = 
+	AnalyzerHelpers::AdjustSimulationTargetSample( largest_sample_requested, sample_rate, mSimulationSampleRateHz );
 
-	while( mSerialSimulationData.GetCurrentSampleNumber() < adjusted_largest_sample_requested )
+	while( mLEDSimulationData.GetCurrentSampleNumber() < adjusted_largest_sample_requested )
 	{
-		CreateSerialByte();
+        WriteReset();
+
+        // six RGB-triple cascade between resets, i.e six discrete LEDs
+        // or two of the 3-LED combined drivers. We should perhaps make
+        // this adjustable
+		for (int t=0; t<6; ++t) {
+			CreateRGBWord();
+		}
 	}
 
-	*simulation_channel = &mSerialSimulationData;
+	*simulation_channel = &mLEDSimulationData;
 	return 1;
 }
 
-void AsyncRgbLedSimulationDataGenerator::CreateSerialByte()
+void AsyncRgbLedSimulationDataGenerator::CreateRGBWord()
 {
-	U32 samples_per_bit = mSimulationSampleRateHz / mSettings->mBitRate;
-
-	U8 byte = mSerialText[ mStringIndex ];
-	mStringIndex++;
-	if( mStringIndex == mSerialText.size() )
-		mStringIndex = 0;
+	const RGBValue rgb = RandomRGBValue();
+	WriteRGBTriple( rgb );
+}
 
-	//we're currenty high
-	//let's move forward a little
-	mSerialSimulationData.Advance( samples_per_bit * 10 );
+void AsyncRgbLedSimulationDataGenerator::WriteRGBTriple( const RGBValue& rgb )
+{
+	U16 values[3];
+	rgb.ConvertToControllerOrder(mSettings->GetColorLayout(), values);
 
-	mSerialSimulationData.Transition();  //low-going edge for start bit
-	mSerialSimulationData.Advance( samples_per_bit );  //add start bit time
+	for (int i=0; i<3; ++i) {
+    	WriteUIntData( values[i], mSettings->BitSize() );
+	}
+}
 
-	U8 mask = 0x1 << 7;
-	for( U32 i=0; i<8; i++ )
-	{
-		if( ( byte & mask ) != 0 )
-			mSerialSimulationData.TransitionIfNeeded( BIT_HIGH );
-		else
-			mSerialSimulationData.TransitionIfNeeded( BIT_LOW );
+void AsyncRgbLedSimulationDataGenerator::WriteReset()
+{
+    // interstitial after last data bit, before the reset goes low
+    mLEDSimulationData.Advance( mClockGenerator.AdvanceByTimeS(1e-5) );
+    mLEDSimulationData.Transition(); // go low
+    // scale by 120% since reset time is the minimum I believe
+    const double resetSec = mClockGenerator.AdvanceByTimeS(mSettings->ResetTimeNSec() * 1.2 * NSEC_TO_SEC);
+    mLEDSimulationData.Advance( resetSec );
+	mLEDSimulationData.Transition(); // go high to end the reset
+}
 
-		mSerialSimulationData.Advance( samples_per_bit );
+void AsyncRgbLedSimulationDataGenerator::WriteUIntData( U32 data, U8 bit_count )
+{
+	U32 mask =  1 << (bit_count - 1);
```

Comment:
just checking here, have you confirmed that all controllers are MSB first? I haven't looked.

Reply:
Also, I AnalyzerHelpers.h contains a class `BitExtractor`, which we should use instead of manually shifting bits using a mask. `BitExtractor` is used for extracting bits from a word. The `DataBuilder` class is used for constructing a word from bits, which you should use in the analysis thread.

Reply:
I think they are all MSB from what I read, but I can check again.

I'll look at the BitExtractor and DataBuilder classes and use them as suggested.